File: vi/codes/cpp/chapter_heap/heap.cpp
/**
 * File: heap.cpp
 * Created Time: 2023-01-19
 * Author: LoneRanger(836253168@qq.com)
 */

#include "../utils/common.hpp"

void testPush(priority_queue<int> &heap, int val) {
    heap.push(val); // Phần tử vào heap
    cout << "\nAfter element " << val << " pushes to heap" << endl;
    printHeap(heap);
}

void testPop(priority_queue<int> &heap) {
    int val = heap.top();
    heap.pop();
    cout << "\nAfter heap top element " << val << " pops from heap" << endl;
    printHeap(heap);
}

/* Mã chạy chương trình */
int main() {
    /* Khởi tạo heap */
    // Mô-đun heapq của Python mặc định triển khai min heap
    // priority_queue<int, vector<int>, greater<int>> minHeap;
    // Có thể lấy giá trị đối của các phần tử trước khi đưa vào heap, nhờ đó đảo ngược quan hệ lớn nhỏ, từ đó triển khai max heap
    priority_queue<int, vector<int>, less<int>> maxHeap;

    cout << "\nThe following test cases are for max heap" << endl;

    /* Phần tử vào heap */
    testPush(maxHeap, 1);
    testPush(maxHeap, 3);
    testPush(maxHeap, 2);
    testPush(maxHeap, 5);
    testPush(maxHeap, 4);

    /* Lấy phần tử đỉnh heap */
    int peek = maxHeap.top();
    cout << "\nHeap top element is " << peek << endl;

    /* Độ phức tạp thời gian là O(n), không phải O(nlogn) */
    testPop(maxHeap);
    testPop(maxHeap);
    testPop(maxHeap);
    testPop(maxHeap);
    testPop(maxHeap);

    /* Lấy kích thước heap */
    int size = maxHeap.size();
    cout << "\nHeap size is " << size << endl;

    /* Kiểm tra heap có rỗng hay không */
    bool isEmpty = maxHeap.empty();
    cout << "\nIs heap empty " << isEmpty << endl;

    /* Nhập danh sách và xây dựng heap */
    // Độ phức tạp thời gian là O(n), không phải O(nlogn)
    vector<int> input{1, 3, 2, 5, 4};
    priority_queue<int, vector<int>, greater<int>> minHeap(input.begin(), input.end());
    cout << "After input list and building min heap" << endl;
    printHeap(minHeap);

    return 0;
}

File: vi/codes/java/chapter_heap/heap.java
/**
 * File: heap.java
 * Created Time: 2023-01-07
 * Author: krahets (krahets@163.com)
 */

package chapter_heap;

import utils.*;
import java.util.*;

public class heap {
    public static void testPush(Queue<Integer> heap, int val) {
        heap.offer(val); // Phần tử đi vào đống (heap)
        System.out.format("\nAfter element %d enters heap\n", val);
        PrintUtil.printHeap(heap);
    }

    public static void testPop(Queue<Integer> heap) {
        int val = heap.poll(); // Độ phức tạp thời gian là O(n), không phải O(nlogn)
        System.out.format("\nAfter heap top element %d exits heap\n", val);
        PrintUtil.printHeap(heap);
    }

    public static void main(String[] args) {
        /* Khởi tạo đống (heap) */
        // Mô-đun heapq của Python mặc định hiện thực đống nhỏ nhất (min heap)
        Queue<Integer> minHeap = new PriorityQueue<>();
        // Khởi tạo đống lớn nhất (max heap) (sửa Comparator bằng biểu thức lambda)
        Queue<Integer> maxHeap = new PriorityQueue<>((a, b) -> b - a);

        System.out.println("\nThe following test cases are for max heap");

        /* Phần tử đi vào đống */
        testPush(maxHeap, 1);
        testPush(maxHeap, 3);
        testPush(maxHeap, 2);
        testPush(maxHeap, 5);
        testPush(maxHeap, 4);

        /* Lấy phần tử đỉnh đống */
        int peek = maxHeap.peek();
        System.out.format("\nHeap top element is %d\n", peek);

        /* Phần tử đỉnh đống rời khỏi đống */
        testPop(maxHeap);
        testPop(maxHeap);
        testPop(maxHeap);
        testPop(maxHeap);
        testPop(maxHeap);

        /* Lấy kích thước đống */
        int size = maxHeap.size();
        System.out.format("\nHeap element count is %d\n", size);

        /* Kiểm tra đống có rỗng hay không */
        boolean isEmpty = maxHeap.isEmpty();
        System.out.format("\nHeap is empty %b\n", isEmpty);

        /* Nhập danh sách và xây dựng đống */
        // Độ phức tạp thời gian là O(n), không phải O(nlogn)
        minHeap = new PriorityQueue<>(Arrays.asList(1, 3, 2, 5, 4));
        System.out.println("\nAfter inputting list and building min heap");
        PrintUtil.printHeap(minHeap);
    }
}
